community/ndp/messaging-v1/src/main/java/org/neo4j/ndp/messaging/v1/NDPMessageFormatV1.java

```diff
@@ -0,0 +1,346 @@
+/*
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.ndp.messaging.v1;
+
+import java.io.IOException;
+import java.util.Map;
+
+import org.neo4j.kernel.api.exceptions.Status;
+import org.neo4j.ndp.messaging.NDPIOException;
+import org.neo4j.ndp.messaging.v1.message.Message;
+import org.neo4j.packstream.PackStream;
+import org.neo4j.ndp.runtime.spi.Record;
+
+import static org.neo4j.ndp.runtime.internal.Neo4jError.codeFromString;
+import static org.neo4j.ndp.runtime.spi.Records.record;
+
+public class NDPMessageFormatV1 implements MessageFormat
+{
+    public static final int VERSION = 1;
+
+    @Override
+    public int version()
+    {
+        return VERSION;
+    }
+
+    public interface MessageTypes
+    {
+        byte MSG_INITIALIZE = 0x01;
+        byte MSG_ACK_FAILURE = 0x0F;
+        byte MSG_RUN = 0x10;
+        byte MSG_DISCARD_ALL = 0x2F;
+        byte MSG_PULL_ALL = 0x3F;
+
+        byte MSG_RECORD = 0x71;
+        byte MSG_SUCCESS = 0x70;
+        byte MSG_IGNORED = 0x7E;
+        byte MSG_FAILURE = 0x7F;
+    }
+
+    static String messageTypeName( int type )
+    {
+        switch( type )
+        {
+        case MessageTypes.MSG_ACK_FAILURE: return "MSG_ACK_FAILURE";
+        case MessageTypes.MSG_RUN:         return "MSG_RUN";
+        case MessageTypes.MSG_DISCARD_ALL: return "MSG_DISCARD_ALL";
+        case MessageTypes.MSG_PULL_ALL:    return "MSG_PULL_ALL";
+        case MessageTypes.MSG_RECORD:      return "MSG_RECORD";
+        case MessageTypes.MSG_SUCCESS:     return "MSG_SUCCESS";
+        case MessageTypes.MSG_IGNORED:     return "MSG_IGNORED";
+        case MessageTypes.MSG_FAILURE:     return "MSG_FAILURE";
+        default: return "0x" + Integer.toHexString(type);
+        }
+    }
+
+    public static class Writer extends CypherPack.Packer implements MessageFormat.Writer
```

Comment:
How come this extends CypherPacker? Users of Writer should not have access to the methods to pack Neo4j data structures - CypherPack.Packer should be a field we delegate to, not a superclass, no? 

Maybe CypherPack could extend the PackStream packers instead, and we use CypherPack here as a delegate, getting access to both the "normal" packstream things as well as the structures. 

Also - could we perhaps call it Neo4jPack instead of CypherPack? It's not dealing with the Cypher type system, it's dealing with the Neo4j type system.